### PR TITLE
Initial deprecations for Labels API

### PR DESCRIPTION
**Pull request by jni**
# References and relevant issues

Initial work towards resolving #6491.

# Description

This deprecates the following attributes in the Labels layer:

- num_colors: `layer.num_colors` becomes `len(layer.colormap)`. `layer.num_colors = n` becomes `layer.colormap = label_colormap(n)`.
- `napari.utils.colormaps.LabelColormap` is deprecated and has been renamed to `napari.utils.colormaps.CyclicLabelColormap`.
- color: `layer.color` becomes `layer.colormap.color_dict`. `layer.color = color_dict` becomes `layer.colormap = DirectLabelColormap(color_dict)`.
- _background_label: `layer._background_label` is now at `layer.colormap.background_value`.
- color_mode: `layer.color_mode` is set by setting the colormap using the corresponding colormap type (`CyclicLabelColormap` or `DirectLabelColormap`).
- `seed_rng`: this was never released, it's just removed.
- `seed`: this was only used for shifting labels around in [0, 1]. It was superseded by #6460 and is deprecated.

### CODE REVIEW

File: napari/layers/labels/labels.py [hunk 1]
```diff
@@ -562,15 +561,34 @@ def _set_colormap(self, colormap):
...
+        warnings.warn(
+            trans._(
+                'Labels.num_colors is deprecated since 0.4.19 and will be '
+                'removed in 0.5.0, please use len(Labels.colormap) - 1 '
+                'instead.',
+                deferred=True,
+            ),
+            FutureWarning,
+            stacklevel=2,
+        )
```
Comment (Czaki) [inline napari/layers/labels/labels.py]:
```suggestion
        warnings.warn(
            trans._(
                'Labels.num_colors is deprecated since 0.4.19 and will be '
                'removed in 0.5.0, please use Labels.colormap.num_colors '
                'instead.',
                deferred=True,
            ),
            FutureWarning,
            stacklevel=2,
        )
```

Comment (Czaki) [reply napari/layers/labels/labels.py]: I do not like this -1. It is technical detail. We should define property (make `_num_unique_colors` property public.

File: napari/layers/labels/labels.py [hunk 2]
```diff
@@ -562,15 +561,34 @@ def _set_colormap(self, colormap):
...
+        return len(self.colormap)
 
     @num_colors.setter
     def num_colors(self, num_colors):
+        warnings.warn(
+            trans._(
+                'Setting Labels.num_colors is deprecated since 0.4.19 and '
+                'will be removed in 0.5.0, please set Labels.colormap '
+                'instead.',
```
Comment (Czaki) [inline napari/layers/labels/labels.py:579]: Maybe point to `napari.utils.colormaps.label_colormap` function?

File: napari/layers/labels/labels.py (hunk 1 above)
Comment (jni) [reply napari/layers/labels/labels.py]: Actually, the -1 is outdated — if you look below, there's no -1. I was just confused because we use different conventions in different places. I think in the future we should harmonize on one definition or the other (I'd say probably including all colors, since then it matches the length of the colorarray) and keep len, which is intuitive. For now I will just update the message to just `len(Label.colormap)` (no -1).

File: napari/layers/labels/labels.py (hunk 2 above)
Comment (jni) [reply napari/layers/labels/labels.py:579]: I think I want to discourage/remove that also in the future — we should be able to use the LabelColormap constructor with a list of colors. Maybe we make a `low_discrepancy_colors` function to generate it and then it's `LabelColormap(low_discrepancy_colors(5))`. What do you think?

File: napari/layers/labels/labels.py (hunk 1 above)
Comment (Czaki) [reply napari/layers/labels/labels.py]:
```python
len(viewer.layers[0].colormap)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[1], line 1
----> 1 len(viewer.layers[0].colormap)

TypeError: object of type 'LabelColormap' has no len()
```

File: napari/layers/labels/labels.py (hunk 2 above)
Comment (Czaki) [reply napari/layers/labels/labels.py:579]: ok

File: napari/layers/labels/labels.py (hunk 1 above)
Comment (jni) [reply napari/layers/labels/labels.py]: Yes you need this PR for that 😉

Comment (Czaki) [reply napari/layers/labels/labels.py]: The current implementation will not work with direct colormap.

File: napari/layers/labels/labels.py (hunk 2 above)
Comment (Czaki) [reply napari/layers/labels/labels.py:579]: And such a function should be part of this PR to allow switching to the new API.

File: napari/layers/labels/labels.py (hunk 1 above)
Comment (jni) [reply napari/layers/labels/labels.py]: Interesting point! I'll think about ways around that... It would be easy to add a len to the direct colormap but then it is a behaviour change. But it's also a bit weird to ask for num_colors while in direct mode. 🤔

Comment (Czaki) [reply napari/layers/labels/labels.py]: And all this problems comes from joining both colormap under one name. 

Maybe t is better to deprecate `colormap` property in Labels layer, and introduce `cyclic_colormap` and `direct_colormap` ones with keeping `color_mode` property?

Comment (jni) [reply napari/layers/labels/labels.py]: Question: would you expect len(a_direct_colormap) to give you the number of colors or the number of keys? Colors is of course much easier to implement and I'm inclined to just do that. But I'm willing to hear counter-arguments. 😊

Comment (Czaki) [reply napari/layers/labels/labels.py]: Keys is much easier to implement (`len(dict)`) than colors (`len(set(tuple(x) for x in dict.values()))`) 

I expect that if I run code that calls `len(Labels.colormap)` it should not require adding `if isinstance(Labels.colormap, LabelsColormap)` to not get some stupid values.

Comment (jni) [reply napari/layers/labels/labels.py]: > Keys is much easier to implement (`len(dict)`) than colors (`len(set(tuple(x) for x in dict.values()))`)

@Czaki no, the point is that actually, it's already implemented for the values, because I added `len` to `Colormap`, not even `LabelsColormapBase`, as `len(self.colors)`. Since you construct `DirectLabelColormap` by making the above set, it already returns the number of colors. Since the answer is accurate (this is in fact the number of colors), I'm inclined to just leave it as-is... For example, if one does a mapping of 1M labels to Viridis discretised to uint8, I think 256 is a reasonable answer.

I guess the question is why people would be reading `layer.num_colors` if they are using direct mode... My guess is that this doesn't happen. (And if it does, we will hear about it as a bug report during the deprecation period (0.4.19 -> 0.5.0).)

Comment (jni) [reply napari/layers/labels/labels.py]: (Hmmm, so do you think I should return `len(self._random_colormap)` in the deprecated implementation? Then people can notice that the results are different when they implement the suggestion.)

Comment (Czaki) [reply napari/layers/labels/labels.py]: > I guess the question is why people would be reading `layer.num_colors` if they are using direct mode... My guess is that this doesn't happen. (And if it does, we will hear about it as a bug report during the deprecation period (0.4.19 -> 0.5.0).)

Because some plugin creators will not think earlier and forget that under colormap it could also be direct one. 

> Since you construct `DirectLabelColormap` by making the above set, it already returns the number of colors

`colors` in `DirecLabelColorap` will have length 1.

https://github.com/napari/napari/blob/d507dd64ed6d96e984fa7fa57f2f232be388b5f7/napari/utils/colormaps/colormap.py#L353-L356

In a direct colormap the colors are stored in `color_dict` attribute. 

It i not my design. I just modified the existing one. 

> do you think I should return `len(self._random_colormap)` in the deprecated implementation

For me, it will be a number of keys. But without strong opinion.

Comment (jni) [reply napari/layers/labels/labels.py]: @Czaki I went with passing through `_num_unique_colors` because it allows unifying a few code paths between direct and auto color modes.

**Comment by jni** [comment]
(I still have a bunch of stuff to do here, sorry for the slowness.)

File: napari/_qt/layer_controls/_tests/test_qt_layer_controls.py [hunk 3]
```diff
@@ -156,6 +167,16 @@ def _create_layer_controls(layer_type_with_data):
         _TRACKS,
         _VECTORS,
     ],
+    ids=[
+        "labels_with_color",
+        "labels",
```
Comment (jni) [inline napari/_qt/layer_controls/_tests/test_qt_layer_controls.py]: maybe:

```suggestion
        "labels_with_direct_colormap",
        "labels_with_auto_colormap",
```
?

File: napari/utils/colormaps/colormap.py [hunk 4]
```diff
@@ -236,13 +243,27 @@ def _selection_as_minimum_dtype(self, dtype: np.dtype) -> int:
...
     use_selection : bool
+        Whether map only selected label.
+        If `True` only selected label will be mapped to not transparent color.
     selection : int
+        The selected label.
+    background_value : int
+        Which value should be treated as a background
+        and mapped to transparent color.
+    interpolation : Literal['zero']
+        required by implementation, please do not set value
```
Comment (jni) [inline napari/utils/colormaps/colormap.py:278]: Can we somehow freeze this value, rather than requiring users to cooperate?

File: napari/utils/colormaps/colormap.py [hunk 5]
```diff
@@ -236,13 +243,27 @@ def _selection_as_minimum_dtype(self, dtype: np.dtype) -> int:
...
+    seed : float
+        seed used for random color generation. Used for reproducibility.
+        It will be removed in the future release.
```
Comment (jni) [inline napari/utils/colormaps/colormap.py:281]: This is only used in the labels layer to store the (deprecated) seed, maybe we should remove it in this PR?

File: napari/utils/colormaps/colormap.py (hunk 4 above)
Comment (aganders3) [reply napari/utils/colormaps/colormap.py:278]:
```diff
diff --git a/napari/utils/colormaps/colormap.py b/napari/utils/colormaps/colormap.py
index d48bcffa3..d43a984de 100644
--- a/napari/utils/colormaps/colormap.py
+++ b/napari/utils/colormaps/colormap.py
@@ -176,7 +176,7 @@ class LabelColormapBase(Colormap):
     background_value: int = 0
     interpolation: Literal[
         ColormapInterpolationMode.ZERO
-    ] = ColormapInterpolationMode.ZERO
+    ] = Field(ColormapInterpolationMode.ZERO, frozen=True)
     _cache_mapping: Dict[Tuple[np.dtype, np.dtype], np.ndarray] = PrivateAttr(
         default={}
     )
```
Does this do what you'd like? With this change, setting the field to anything else would raise a `ValidationError`:
```
ValidationError: 1 validation error for LabelColormap
interpolation
  unexpected value; permitted: <ColormapInterpolationMode.ZERO: 'zero'> (type=value_error.const; given=linear; permitted=(<ColormapInterpolationMode.ZERO: 'zero'>,))
  ```

File: napari/utils/colormaps/colormap.py (hunk 5 above)
Comment (jni) [reply napari/utils/colormaps/colormap.py:281]: GH code search returns 0 hits of user code using `colormap.seed`. So I would vote for removing this directly.

Comment (Czaki) [reply napari/utils/colormaps/colormap.py:281]: This seed is a parameter of `label_colormap` function also that is part of the public API for creating colormap. 

The general story behind this change is that until this PR, the colormaps were not part of public API documentation. I try to set up [`autodoc-pydantic`](https://pypi.org/project/autodoc-pydantic/), but it do not work out of the box. So I decided to document missing parameters. 

As we miss many docstrings in our codebase, I prefer to clearly write that argument is deprecated, than left it undocumented.

Comment (jni) [reply napari/utils/colormaps/colormap.py:281]: I'm just saying it's not used even after it is stored. So maybe we just remove it from the class, not leave it undocumented. Yes it is a public API but I don't think anyone is using it (and if they are they have plenty of recourse to fix in their own code). But, maybe we can do that in 0.5.0, is that what you would prefer?

File: napari/layers/labels/labels.py [hunk 6]
```diff
@@ -287,22 +288,20 @@ def __init__(
...
         name=None,
-        num_colors=49,
         opacity=0.7,
         plane=None,
         properties=None,
         projection_mode='none',
         rendering='iso_categorical',
         rotate=None,
         scale=None,
-        seed_rng=None,
```
Comment (brisvag) [inline napari/layers/labels/labels.py:303]: is this supposed to be gone like this with no warning?

File: napari/layers/labels/labels.py [hunk 7]
```diff
@@ -311,14 +310,14 @@ def __init__(
             name = magic_name(data)
 
         self._seed = 0.5
-        self._seed_rng: Optional[int] = seed_rng
-        self._background_label = 0
-        self._num_colors = num_colors
         self._random_colormap = label_colormap(
-            self.num_colors, self.seed, self._background_label
+            49, self._seed, background_value=0
```
Comment (brisvag) [inline napari/layers/labels/labels.py:314]: maybe set this `49` to a variable and/or add a comment so we don't lose track of the origin and reason.

**Review by brisvag — COMMENTED** [review]
Generally all good, left a couple comments!

File: napari/layers/labels/labels.py (hunk 6 above)
Comment (Czaki) [reply napari/layers/labels/labels.py:303]: Yes. As it was introduced after 0.4.18 release.

**Comment by aganders3** [comment]
I'm having a problem with this PR but it might be user error.

If I create a new labels layer and try to change it to "direct" color mode, nothing happens and the color mode switches back to "auto". Is this working as intended? I would expect "direct" to be absent or grayed out in the dropdown if it's not going to change.

Then I gathered I need to set a color dict before changing the mode. I get warnings to use `layer.colormap.color_dict` instead of setting `layer.color`, but this doesn't work if the layer is not in direct mode (`AttributeError: 'LabelColormap' object has no attribute 'color_dict'`).

**Comment by jni** [comment]
> If I create a new labels layer and try to change it to "direct" color mode, nothing happens and the color mode switches back to "auto". Is this working as intended? 

This is almost certainly the same behaviour as in main, and imho all the more reason to deprecate this forthwith. Actually this PR was originally going to have a colormap selector in place of the color mode but we decided to leave it till later.

> I would expect "direct" to be absent or grayed out in the dropdown if it's not going to change.

True enough, but since we are just going to get rid of the thing soon, I suggest putting this in the wontfix bucket. (We can debate about whether we want to change the UI before 0.4.19.)

> I get warnings to use layer.colormap.color_dict instead of setting layer.color, but this doesn't work if the layer is not in direct mode (AttributeError: 'LabelColormap' object has no attribute 'color_dict').

This is a good catch: the warning works for *reading* the attribute, but not for setting it. I'll update that. 👍

File: napari/layers/labels/labels.py (hunk 7 above)
Comment (jni) [reply napari/layers/labels/labels.py:314]: Just gonna add a comment. I think this is minor and may change in the future (e.g. with a default colormap name) so I don't think it warrants its own variable. 😅

**Comment by jni** [comment]
> > I get warnings to use layer.colormap.color_dict instead of setting layer.color, but this doesn't work if the layer is not in direct mode (AttributeError: 'LabelColormap' object has no attribute 'color_dict').
> 
> This is a good catch: the warning works for reading the attribute, but not for setting it. I'll update that. 👍

Actually, the issue is when reading, the colormap *can* be of the wrong type. I've updated the warning so that it specifically mentions that this can only be used when the colormap is of a particular type.

File: napari/layers/labels/labels.py [hunk 8]
```diff
@@ -972,14 +1006,6 @@ def _calculate_contour(
...
         return sliced_labels[delta_slice]
 
-    def _get_cache_dtype(self, raw_dtype: np.dtype) -> np.dtype:
-        if self.color_mode == LabelColorMode.DIRECT:
-            return _texture_dtype(
-                self._direct_colormap._num_unique_colors + 2,
-                raw_dtype,
-            )
-        return _texture_dtype(self.num_colors, raw_dtype)
-
```
Comment (Czaki) [inline napari/layers/labels/labels.py]: this removal leads to current crash

**Comment by aganders3** [comment]
> This is almost certainly the same behaviour as in main, and imho all the more reason to deprecate this forthwith.

For what it's worth I can't reproduce this on main. Instead it churns for a beat, but then the mode is indeed changed - not particularly helpful because it also makes the labels all transparent black. However, I understand if this is an acceptable part of the deprecation story.

The difference in behavior is because this code is now on the path when using the combobox, where on main it is avoided by setting `color_mode` directly. Maybe there should also be a warning here?
https://github.com/napari/napari/blob/7a052349af587e36eb4c42326154e559c1a655dd/napari/layers/labels/labels.py#L541-L554

Here's one idea for disabling the combo box - no worries if you want to keep that out of this PR. Maybe this requires a tooltip explaining why the box is disabled.
```diff
diff --git a/napari/_qt/layer_controls/qt_labels_controls.py b/napari/_qt/layer_controls/qt_labels_controls.py
index f10da2c0e..b78754ac5 100644
--- a/napari/_qt/layer_controls/qt_labels_controls.py
+++ b/napari/_qt/layer_controls/qt_labels_controls.py
@@ -311,6 +311,9 @@ class QtLabelsControls(QtLayerControls):
             self.layer.colormap = self.layer._direct_colormap
 
     def _on_colormap_change(self):
+        self.colorModeComboBox.setEnabled(
+            not self.layer._is_default_colors(self.layer._direct_colormap.color_dict)
+        )
         if isinstance(self.layer.colormap, LabelColormap):
             self.colorModeComboBox.setCurrentIndex(
                 self.colorModeComboBox.findData(LabelColorMode.AUTO.value)
```

File: napari/layers/labels/labels.py [hunk 9]
```diff
@@ -1001,6 +1007,9 @@ def _calculate_contour(
         )
         return sliced_labels[delta_slice]
 
+    def _get_cache_dtype(self, raw_dtype: np.dtype) -> np.dtype:
+        return _texture_dtype(len(self.colormap), raw_dtype)
+
```
Comment (brisvag) [inline napari/layers/labels/labels.py]: is this temporary to fix a bug and then should go away? If so, maybe put a comment so it's easy to find when we complete the deprecation.

**Review by brisvag — APPROVED** [review]
Still looks good in general :)

**Review by brisvag — APPROVED** [review]
Tok a higher level look at the API changes, and I agree with everything. One note: does `new_colormap` need to exist on the public api? It seems like it serves similar purpose as the old `seed`, maybe it should be private or exposed on the colormap itself?

**Comment by Czaki** [comment]
> One note: does `new_colormap` need to exist on the public api?

The name is wrong. I should be something like "shuffle_colormap" 
The current difference is that there is no seed attribute of layer, that is a huge improvement. 
It also allows to "shuffle" any colormap, not only one class of them.

**Comment by brisvag** [comment]
Then I propose to rename to `shuffle_colormap` or `shuffle_colors` and deprecate together with the rest... and/or alternatively remove it from the layer and have it just on the colormap directly.

File: napari/layers/labels/labels.py [hunk 10]
```diff
@@ -555,26 +552,45 @@ def _set_colormap(self, colormap):
...
+        self.refresh()
 
     @property
     def num_colors(self):
         """int: Number of unique colors to use in colormap."""
-        return self._num_colors
+        warnings.warn(
+            trans._(
+                'Labels.num_colors is deprecated since 0.4.19 and will be '
+                'removed in 0.5.0, please use len(Labels.colormap) '
```
Comment (psobolewskiPhD) [inline napari/layers/labels/labels.py:564]: I'm all for removing `num_colors`, but kind hate the alternative being `len(...)`, because of the lack of suggest/complete using IDE/ipython/console... 
What about having `num_colors` on the Colormap object? So one could do: `Labels.colormap.num_colors`

File: napari/layers/labels/labels.py [hunk 11]
```diff
@@ -97,9 +97,9 @@ class Labels(_ImageBase):
     cache : bool
         Whether slices of out-of-core datasets should be cached upon retrieval.
         Currently, this only applies to dask arrays.
-    color : dict of int or None to str or array
-        Custom label to color mapping. Values must be valid color names or RGBA
-        arrays. None is used when no color is specified for a label.
+    colormap : CycleLabelColormap or DirectLabelColormap or None
```
Comment (psobolewskiPhD) [inline napari/layers/labels/labels.py]: I like this change from color to colormap, makes it more intuitive.
Any thought to `CyclicLabelColormap` vs `CycleLabelColormap` ? I prefer the first, but not strongly.

File: napari/layers/labels/labels.py [hunk 12]
```diff
@@ -189,9 +186,11 @@ class Labels(_ImageBase):
         Custom label to color mapping. Values must be valid color names or RGBA
         arrays. While there is no limit to the number of custom labels, the
         the layer will render incorrectly if they map to more than 1024 distinct
-        colors.
+        colors. DEPRECATED: set ``colormap`` directly, using
+        `napari.utils.colormaps.direct_colormap`.
```
Comment (psobolewskiPhD) [inline napari/layers/labels/labels.py]:
```suggestion
        `napari.utils.colormaps.direct_colormap` or `napari.utils.colormaps.label_colormap`
```

File: napari/layers/labels/labels.py [hunk 13]
```diff
@@ -279,6 +278,8 @@ class Labels(_ImageBase):
 
     _history_limit = 100
 
+    @deprecated_constructor_arg_by_attr("color")
```
Comment (psobolewskiPhD) [inline napari/layers/labels/labels.py:279]: I tried to pass `color=dict_of_colors` to `add_labels` and got
`TypeError: add_labels() got an unexpected keyword argument 'color'`
I was expecting to get a deprecation warning...

Comment (psobolewskiPhD) [reply napari/layers/labels/labels.py:279]: There are also two uses of `color=` in:
`benchmarks/benchmark_labels_layer.py`
line 101 and 135

File: napari/layers/labels/labels.py [hunk 14]
```diff
@@ -785,17 +796,41 @@ def color_mode(self):
 
         DIRECT allows color of each label to be set directly by a color dict.
         """
+        warnings.warn(
+            trans._(
+                'Labels.color_mode is deprecated since 0.4.19 and will be '
+                'removed in 0.5.0. Please check type(Labels.colormap) '
+                'instead. napari.utils.colormaps.LabelColormap corresponds to '
```
Comment (psobolewskiPhD) [inline napari/layers/labels/labels.py]:
```suggestion
                'instead. napari.utils.colormaps.CycleLabelColormap corresponds to '
```
Right? LabelColormap is the thing being deprecated?

File: napari/layers/labels/labels.py [hunk 15]
```diff
@@ -785,17 +796,41 @@ def color_mode(self):
...
+        )
         return str(self._color_mode)
 
     @color_mode.setter
     def color_mode(self, color_mode: Union[str, LabelColorMode]):
+        warnings.warn(
+            trans._(
+                'Labels.color_mode is deprecated since 0.4.19 and will be '
+                'removed in 0.5.0. Please set Labels.colormap instead, to an'
+                'instance of napari.utils.colormaps.LabelColormap for "auto"'
```
Comment (psobolewskiPhD) [inline napari/layers/labels/labels.py]:
```suggestion
                'instance of napari.utils.colormaps.CycleLabelColormap for "auto"'
```

File: examples/add_labels_with_features.py [hunk 16]
```diff
@@ -56,7 +58,7 @@
     label_image,
     name='segmentation',
     features=label_features,
-    color=color,
+    colormap=DirectLabelColormap(color_dict=colors),
```
Comment (psobolewskiPhD) [inline examples/add_labels_with_features.py:61]: This example ends up looking different than in the gallery.
With this PR I see:
image

with main:
image

I'm not which is more correct, because I'm not sure if I get this example. I'd expect the large coins to get one color and the small ones to get a different color.

Comment (psobolewskiPhD) [reply examples/add_labels_with_features.py:61]: I guess if None is to be magenta, then the new way is technically correct?

File: napari/components/_tests/test_prune_kwargs.py [hunk 17]
```diff
@@ -29,7 +28,6 @@
         'labels',
         {
             'scale': (0.75, 1),
-            'num_colors': 10,
```
Comment (psobolewskiPhD) [inline napari/components/_tests/test_prune_kwargs.py:32]: The docstring of this function in `viewer_model.py` also uses `num_colors` in the example.

File: napari/utils/colormaps/colormap.py [hunk 18]
```diff
@@ -331,17 +353,29 @@ def shuffle(self, seed: int):
...
 
     Attributes
     ----------
     color_dict: dict from int to (3,) or (4,) array
         The dictionary mapping labels to colors.
+
     use_selection : bool
-        Whether to color using the selected label.
+        Whether map only selected label.
+        If `True` only selected label will be mapped to not transparent color..
```
Comment (psobolewskiPhD) [inline napari/utils/colormaps/colormap.py]:
```suggestion
        Whether to map only the selected label to a color.
        If `True` only selected label will be not transparent.
```

**Review by psobolewskiPhD — CHANGES_REQUESTED** [review]
Sorry for the delay. I like the concept overall, it's more intuitive--to me at least. 
I think I caught a few missed places for deprecations? 
Also I think having `colormap.num_colors` or such would be nicer than using `len(colormap)`

Re: naming
I'm OK with the current or using DictColormap and CyclicColormap to make them more generic.

File: examples/add_labels_with_features.py (hunk 16 above)
Comment (Czaki) [reply examples/add_labels_with_features.py:61]: Yes. The old behavior was adding transparent color for the background. There is a need to add `0: 'transparent'` to color dict.

Comment (jni) [reply examples/add_labels_with_features.py:61]: I think we should maybe add the background = 0 behaviour as default to the DirectLabelColormap to preserve the old behaviour (and this is the most intuitive behaviour for our users, I think). Then it could be disabled by passing `background_value=None` to the constructor. Thoughts?

Comment (Czaki) [reply examples/add_labels_with_features.py:61]: I do not like it. It complicates API. But we could to this with `direct_colormap` helper function https://github.com/napari/napari/blob/3c58ab41e1a1772f133852a6c6c4adf7e209a576/napari/utils/colormaps/colormap_utils.py#L567-L585

File: napari/layers/labels/labels.py (hunk 13 above)
Comment (Czaki) [reply napari/layers/labels/labels.py:279]: Will fix the problem with `add_labels tomorrow

File: examples/add_labels_with_features.py (hunk 16 above)
Comment (jni) [reply examples/add_labels_with_features.py:61]: Ah, I see that the background value *is* 0 by default. Well, I think by default the background should always be fully transparent — that is, if there is not a key for the background value, we add it. @Czaki?

Comment (jni) [reply examples/add_labels_with_features.py:61]: I've implemented this in ddbe9c8 (with a slight tweak to the docstring in 8709db7)

File: napari/components/_tests/test_prune_kwargs.py (hunk 17 above)
Comment (jni) [reply napari/components/_tests/test_prune_kwargs.py:32]: Fixed in df05b49

File: napari/layers/labels/labels.py (hunk 10 above)
Comment (jni) [reply napari/layers/labels/labels.py:564]: @psobolewskiPhD maybe we can discuss this in a follow-up issue or PR? I am of the opposite opinion, because we have way too many attributes in our objects — frequently when using napari I do `obj.<TAB>` and then I get an enormous list in which it is really hard to find what I'm looking for. I'm not sure whether colormaps fall in this category, but the point is that *eventually* all objects land in this category if you keep expanding the API for every thing that you want to know about. And, it's also real easy to keep adding things, but hard to remove them, so, especially while we are working to simplify our APIs, for every attribute we add to a class, I want to have a nice, long discussion about it. 😅 For standard Python things like "how many things are in this collection", I prefer to use `len()`. I realise that not everyone would think to do this, which is why I'm happy to contemplate in a follow-up issue, but since adding to the API is easy and not time-sensitive, I suggest we leave the discussion until post-0.5.

File: napari/layers/labels/labels.py (hunk 11 above)
Comment (jni) [reply napari/layers/labels/labels.py]: I like that now the two class names are the same length. 😊 Done in 6f340e8

File: napari/layers/labels/labels.py (hunk 13 above)
Comment (jni) [reply napari/layers/labels/labels.py:279]: @Czaki happy to take this on if you point me to how I should fix this...

File: napari/layers/labels/labels.py (hunk 12 above)
Comment (jni) [reply napari/layers/labels/labels.py]: No, here we are replacing setting color, which is the direct_colormap. For num_colors, users would use label_colormap.

Comment (jni) [reply napari/layers/labels/labels.py]: (However, I've changed the pointer to the class rather than the helper function.)

**Comment by jni** [comment]
@Czaki there's a [weird error](https://github.com/napari/napari/actions/runs/7609291570/job/20720239205?pr=6542#step:11:418) in the runner (not sure if it's from this branch):

```pytb
  napari/_tests/test_pytest_plugin.py Process pystack_monitor:
  Traceback (most recent call last):
    File "/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/multiprocessing/process.py", line 315, in _bootstrap
      self.run()
    File "/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/multiprocessing/process.py", line 108, in run
      self._target(*self._args, **self._kwargs)
    File "/home/runner/work/napari/napari/.tox/py38-linux-pyqt5-no_cov/lib/python3.8/site-packages/pytest_pystack/_monitor_process.py", line 57, in _run_monitor
      raise Exception(
  Exception: new test should not start before previous test finished
```

However, weirdly, it doesn't seem to cause the tests to fail. (But it is reproducible...)

However, [this](https://github.com/napari/napari/actions/runs/7609291570/job/20720239205?pr=6542#step:11:452) is a real failure:

```pytb
  ____________________________ test_signature[Labels] ____________________________
  
  layer = <class 'napari.layers.labels.labels.Labels'>
  
      @pytest.mark.parametrize('layer', layers, ids=lambda layer: layer.__name__)
      def test_signature(layer):
          name = layer.__name__
      
          method = getattr(Viewer, f'add_{camel_to_snake(name)}')
      
          class_parameters = dict(inspect.signature(layer.__init__).parameters)
          method_parameters = dict(inspect.signature(method).parameters)
      
          fail_msg = f"signatures don't match for class {name}"
          if name == 'Image':
              # If Image just test that class params appear in method
              for class_param in class_parameters:
                  assert class_param in method_parameters, fail_msg
          else:
  >           assert class_parameters == method_parameters, fail_msg
  E           AssertionError: signatures don't match for class Labels
  E           assert {'affine': <P...p=None">, ...} == {'affine': <P...r=None">, ...}
  E             Omitting 22 identical items, use -vv to show
  E             Right contains 3 more items:
  E             {'color': <Parameter "color=None">,
  E              'num_colors': <Parameter "num_colors=None">,
  E              'seed': <Parameter "seed=None">}
  E             Full diff:
  E               {
  E                'affine': <Parameter "affine=None">,
  E                'blending': <Parameter "blending='translucent'">,
  E                'cache': <Parameter "cache=True">,
  E             -  'color': <Parameter "color=None">,
  E                'colormap': <Parameter "colormap=None">,
  E                'data': <Parameter "data">,
  E                'depiction': <Parameter "depiction='volume'">,
  E                'experimental_clipping_planes': <Parameter "experimental_clipping_planes=None">,
  E                'features': <Parameter "features=None">,
  E                'metadata': <Parameter "metadata=None">,
  E                'multiscale': <Parameter "multiscale=None">,
  E                'name': <Parameter "name=None">,
  E             -  'num_colors': <Parameter "num_colors=None">,
  E                'opacity': <Parameter "opacity=0.7">,
  E                'plane': <Parameter "plane=None">,
  E                'projection_mode': <Parameter "projection_mode='none'">,
  E                'properties': <Parameter "properties=None">,
  E                'rendering': <Parameter "rendering='iso_categorical'">,
  E                'rotate': <Parameter "rotate=None">,
  E                'scale': <Parameter "scale=None">,
  E             -  'seed': <Parameter "seed=None">,
  E                'self': <Parameter "self">,
  E                'shear': <Parameter "shear=None">,
  E                'translate': <Parameter "translate=None">,
  E                'visible': <Parameter "visible=True">,
  E               }
  
  napari/_tests/test_view_layers.py:124: AssertionError
```

Maybe that test can be updated to ignore deprecated parameters? I'm not sure.

File: napari/utils/_register.py [hunk 19]
```diff
@@ -76,36 +93,3 @@ def create_func(cls, name=None, doc=None, filename: str = '<string>'):
...
-        Docstring to use in the method.
-        If None, is autogenerated from the existing docstring.
-    """
-    if cls is not None:
-        return _register(cls, name=name, doc=doc)
-
-    def inner(cls):
-        return _register(cls, name=name, doc=doc)
-
-    return inner
```
Comment (jni) [inline napari/utils/_register.py:111]: Why are these deleted?

Comment (Czaki) [reply napari/utils/_register.py:111]: Because it is unused.

Comment (jni) [reply napari/utils/_register.py:111]: Seems like it should be in a separate PR but ok. 👍

File: napari/utils/colormaps/colormap.py [hunk 20]
```diff
@@ -350,17 +372,29 @@ def shuffle(self, seed: int):
...
+
     use_selection : bool
-        Whether to color using the selected label.
+        Whether to map only the selected label to a color.
+        If `True` only selected label will be not transparent.
     selection : int
         The selected label.
+    colors : ColorArray
+        Exist because of implementation details. Please do not use it.
     """
```
Comment (Czaki) [inline napari/utils/colormaps/colormap.py:398]: As background is used for validation of color_dict then it should be documented.

File: napari/layers/labels/_tests/test_labels.py [hunk 21]
```diff
@@ -263,26 +285,28 @@ def test_seed():
...
+        assert layer.num_colors == 60
 
-    with pytest.raises(
-        ValueError, match=r".*Only up to 2\*\*16=65535 colors are supported"
-    ):
-        layer.num_colors = 2**17
+        with pytest.raises(
+            ValueError, match=r".*Only up to 2\*\*16=65535 colors"
+        ):
+            layer.num_colors = 2**17
```
Comment (Czaki) [inline napari/layers/labels/_tests/test_labels.py]: only line triggering warning should be under `pytest.warns`. So it means that there should be multiple `with pytest.warns`

File: napari/layers/labels/labels.py (hunk 2 above)
Comment (Czaki) [reply napari/layers/labels/labels.py:579]: So what here?

File: napari/utils/migrations.py [hunk 22]
```diff
@@ -165,3 +170,44 @@ def _wrapper(*args, **kwargs):
...
+
+    class _OldClass(new_class):
+        def __new__(cls, *args, **kwargs):
+            warnings.warn(msg, FutureWarning, stacklevel=2)
+            if super().__new__ is object.__new__:
+                return super().__new__(cls)
+            return super().__new__(cls, *args, **kwargs)
+
+        def __init_subclass__(cls, **kwargs):
+            warnings.warn(msg, FutureWarning, stacklevel=2)
```
Comment (Czaki) [inline napari/utils/migrations.py:206]: I think that these lines should be tested.

File: napari/layers/labels/labels.py (hunk 10 above)
Comment (psobolewskiPhD) [reply napari/layers/labels/labels.py:564]: fair enough

**Review by psobolewskiPhD — APPROVED** [review]
